src/evaluators/JSXElement.js

```diff
@@ -221,90 +223,153 @@ function evaluateJSXAttributes(
   strictCode: boolean,
   env: LexicalEnvironment,
   realm: Realm
-): { attributes: Map<string, Value>, children: ArrayValue | Value | null } {
-  let attributes = new Map();
+): { key: Value, props: ObjectValue | AbstractObjectValue, ref: Value } {
+  let props = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
   let children = evaluateJSXChildren(astChildren, strictCode, env, realm);
+  let key = realm.intrinsics.null;
+  let ref = realm.intrinsics.null;
   let defaultProps = getDefaultProps(elementType, env, realm);
+  // to be used if we need to create an abstract props
+  let abstractPropsArgs = [];
+
+  let containsAbstractSpreadAttribute = false;
+  let containsAbstractKey = false;
+  let containsAbstractRef = false;
+  let spreadAttributesCount = 0;
+  let spreadAttributesWithInitialPropsHintCount = 0;
+  let attributesAssigned = 0;
+
+  const setProp = (name: string, value: Value, isDefaultProps: boolean): void => {
+    if (name === "children") {
+      if (isDefaultProps && children !== null) {
+        return;
+      }
+      invariant(props instanceof ObjectValue);
+      Properties.Set(realm, props, "children", value, true);
+    } else if (name === "key" && value !== realm.intrinsics.null) {
+      if (value instanceof AbstractValue) {
+        containsAbstractKey = true;
+      } else if (containsAbstractKey) {
+        containsAbstractKey = false;
+      }
+      key = computeBinary(realm, "+", realm.intrinsics.emptyString, value);
+    } else if (name === "ref") {
+      if (value instanceof AbstractValue) {
+        containsAbstractRef = true;
+      } else if (containsAbstractRef) {
+        containsAbstractRef = false;
+      }
+      ref = value;
+    } else if (name !== "__self" && name !== "__source") {
+      invariant(props instanceof ObjectValue);
```

Comment:
Can you clarify why invariant is useful to run on every prop assignment? Do you expect this to change? Can we run it just once?

Reply:
Flow complains if we don't. Annoying, I know. I've already told the Flow team a while ago about this and it's on their roadmap.